# Типы данных

## Числовые типы данных

### int

In [64]:
x = 5

print(x)
print(type(x))

5
<class 'int'>


In [65]:
a = 4 + 5
b = 4 * 5
c = 5 / 4

print(a, b, c)

9 20 1.25


In [66]:
x = 5 * 1000000 * 1000000 * 1000000 * 1000000 + 1

print(x)
print(type(x))

5000000000000000000000001
<class 'int'>


### float

In [67]:
y = 5.7

print(y)
print(type(y))

5.7
<class 'float'>


In [68]:
a = 4.2 + 5.1
b = 4.2 * 5.1
c = 5.0 / 4.0

print(a, b, c)

9.3 21.419999999999998 1.25


In [69]:
a = 5
b = 4

print(float(a) / float(b))

1.25


In [70]:
print(5 / 4)

1.25


## Логические типы данных

### bool

In [71]:
a = True
b = False

print(a, type(a))
print(b, type(b))

True <class 'bool'>
False <class 'bool'>


In [72]:
print(a + b)
print(a + a)
print(b + b)

1
2
0


In [73]:
print(int(a), int(b))

1 0


In [74]:
print(True and False)
print(True and True)
print(False and False)

False
True
False


In [75]:
print(True or False)
print(True or True)
print(False or False)

True
True
False


### None

Специальный тип который означает отсутствие значения.

In [76]:
z = None

print(z)
print(type(z))

None
<class 'NoneType'>


## Строковые типы данных

### str

In [77]:
x = 'abc'

print(x)
print(type(x))

abc
<class 'str'>


In [78]:
a = 'Ivan'
b = 'Ivanov'

s = a + ' ' + b

print(s)

Ivan Ivanov


In [79]:
print(a.upper(), b.lower())

IVAN ivanov


In [80]:
print(len(a))

4


In [81]:
print(bool(a))
print(bool(''))

True
False


In [82]:
print(int(a))

ValueError: invalid literal for int() with base 10: 'Ivan'

In [83]:
print(a)         # Строка целиком
print(a[0])      # Первый символ строки
print(a[1])      # Второй символ строки
print(a[0:3])    # С нулевого по второй! символ строки
print(a[0:4:2])  # Третий параметр - шаг. Каждый второй символ от среза строки [0:4]

Ivan
I
v
Iva
Ia


In [84]:
splitted_text = 'Ivanov Ivan Ivanovich'.split(' ')

print(splitted_text)
print(type(splitted_text))

['Ivanov', 'Ivan', 'Ivanovich']
<class 'list'>


### bytes

Тип `bytes` появился в Python 3.x. Это переименованный тип `str` из Python 2.x:

````
|Python 2.x | Python 3.x |
|-----------|------------|
|str        | bytes      |
|unicode    | str        |
````

Кодирование строки в байты с помощью таблицы UTF-8:

In [96]:
b = 'пайтон3'.encode('utf-8')

print(b)
print(type(b))

b'\xd0\xbf\xd0\xb0\xd0\xb9\xd1\x82\xd0\xbe\xd0\xbd3'
<class 'bytes'>


Раскодирование байт в символы с помощью таблицы UTF-8

In [97]:
s = b.decode('utf-8')

print(s)
print(type(s))

пайтон3
<class 'str'>


Кодировать строку в байты, также можно с помощью специального префикса 'b', но только для символов из таблицы ASCII:

In [98]:
b = b'python3'

print(b)
print(type(b))

b'python3'
<class 'bytes'>


## Массивы

### list

Список. Изменяемый тип данных.

In [139]:
client = ['Иван', ' Иванович', 50000, 12, True]

print(client)
print(type(client))

['Иван', ' Иванович', 50000, 12, True]
<class 'list'>


Доступ к элементам списка через индекс элемента:

In [140]:
client[0]

'Иван'

In [141]:
client[2]

50000

In [142]:
client[4]

True

Мы можем использовать срезы (slices):

In [143]:
client[0:2]

['Иван', ' Иванович']

In [144]:
client[:2]

['Иван', ' Иванович']

In [145]:
client[-3:]

[50000, 12, True]

In [146]:
client[:-1]

['Иван', ' Иванович', 50000, 12]

Мы можем добавлять, удалять элементы:

In [147]:
client.append(2017)

print(client)

['Иван', ' Иванович', 50000, 12, True, 2017]


In [149]:
client.remove(2017)

print(client)

['Иван', ' Иванович', 50000, 12, True]


In [150]:
v = client.pop()

print(v)
print(client)

True
['Иван', ' Иванович', 50000, 12]


### tuple

Кортеж. Неизменяемый тип данных. Невозможно добавлять/удалять элементы массива (кортежа).

In [151]:
client = ('Иван', ' Иванович', 50000, 12, True)

print(client)
print(type(client))

('Иван', ' Иванович', 50000, 12, True)
<class 'tuple'>


In [152]:
client[:2]

('Иван', ' Иванович')

## Множества и словари

### set

Изменяемый список из уникальных элементов

In [160]:
names = {'Ivan', 'Petr', 'Konstantin'}

print(names)
print(type(names))

{'Konstantin', 'Ivan', 'Petr'}
<class 'set'>


In [161]:
print('Petr' in names)

True


In [162]:
print('Mikhail' in names)

False


In [163]:
names.add('Mikhail')

print(names)

{'Mikhail', 'Konstantin', 'Ivan', 'Petr'}


Попытаемся добавить уже существующий элемент и увидим, что он не добавится:

In [164]:
names.add('Mikhail')

print(names)

{'Mikhail', 'Konstantin', 'Ivan', 'Petr'}


In [165]:
names.remove('Mikhail')

print(names)

{'Konstantin', 'Ivan', 'Petr'}


Добавить список как элемент множества не получится, так как `list` не является хешируемым объектом, т.е. элемент должен быть неизменяемым:

In [166]:
names.add(['Vladimir', 'Vasya'])

print(names)

TypeError: unhashable type: 'list'

Добавить `tuple` в качестве элемента `set` получится, так как он является хешируемым или неизменяемым типом данных:

In [168]:
names.add(('Vladimir', 'Vasya'))

print(names)

{'Petr', 'Konstantin', 'Ivan', ('Vladimir', 'Vasya')}


Можно задаться вопросом: А зачем нужен `set`, если есть уже `list` в котором можно было поддержать уникальность элементов? Ответ прост, реализация `set` гораздо быстрее выполняет поиск элемента, чем если бы реализовали это в `list`. Убедимся в этом на практике:

In [188]:
a = list(range(1000000))
b = set(range(1000000))

Убедимся, что в наших массивах достаточное количество элементов для теста:

In [189]:
print(a[:5], a[-5:])

[0, 1, 2, 3, 4] [999995, 999996, 999997, 999998, 999999]


In [192]:
%%time

print(999999 in a)

True
CPU times: user 12.5 ms, sys: 0 ns, total: 12.5 ms
Wall time: 12.6 ms


In [193]:
%%time

print(999999 in b)

True
CPU times: user 367 µs, sys: 293 µs, total: 660 µs
Wall time: 665 µs


Как видим, в `list` поиск элемента занял 12.5 миллисекунд, а по `set` поиск выполнился за 665 наносекунд!

### dict

Изменяемый тип данных.

In [194]:
d = dict()

print(d)
print(type(d))

{}
<class 'dict'>


In [196]:
d['I']  = 1  # Создает новый элемент
d['am'] = 1
d['I'] += 1  # Изменяет значение существующего элемента

print(d)

{'I': 2, 'am': 1}


In [197]:
print(d['I'])

2


In [198]:
d = {'abc':3.4, 5:7.8, '123':None}

print(d)
print(type(d))

{'abc': 3.4, 5: 7.8, '123': None}
<class 'dict'>


В словаре в качестве ключа может быть кортеж (tuple), а значениями любой тип данных:

In [199]:
d[(1,2,5)] = [4,5,7]

print(d)

{'abc': 3.4, 5: 7.8, '123': None, (1, 2, 5): [4, 5, 7]}


А вот список не может быть в качестве ключа, так как `list` не является хешируемым объектом:

In [200]:
d[[1,2]] = [4,5]

print(d)

TypeError: unhashable type: 'list'